### PR TITLE
[SPARK-20680][SQL][FOLLOW-UP] Make NullType.simpleString as void to support hive

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

For better compatible with Hive in `create view` with null.

Change `NullType.simpleString` to `void`, then Hive can parse `void` to it's own type. 

### Why are the changes needed?

Currently, we support `create view v as select null` with in-memory but failed in hive. We try to forbid this in [#29152](https://github.com/apache/spark/pull/29152), however `df.createTempView` is already a use case of `NullType` with in-memory. So it seems better to keep this behavior consistent with hive.

Some error msg with `create view v as select null`.
```
org.apache.spark.SparkException: Cannot recognize hive type string: null, column: NULL
	at org.apache.spark.sql.hive.client.HiveClientImpl$.getSparkSQLDataType(HiveClientImpl.scala:999)
	at org.apache.spark.sql.hive.client.HiveClientImpl$.$anonfun$verifyColumnDataType$1(HiveClientImpl.scala:1021)
	at scala.collection.Iterator.foreach(Iterator.scala:941)
	at scala.collection.Iterator.foreach$(Iterator.scala:941)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at org.apache.spark.sql.types.StructType.foreach(StructType.scala:102)
```

### Does this PR introduce _any_ user-facing change?

Yes, user can create view with null type in Hive.

### How was this patch tested?

New UT with just Hive. `in-memory` test is already exists.

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 1]
```diff
@@ -984,7 +984,7 @@ private[hive] object HiveClientImpl extends Logging {
     val typeString = if (c.metadata.contains(HIVE_TYPE_STRING)) {
       c.metadata.getString(HIVE_TYPE_STRING)
     } else {
-      c.dataType.catalogString
+      c.dataType.sql
```
Comment (ulysses-you) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:987]: we should use `.sql` to hive type.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/NullType.scala [hunk 2]
```diff
@@ -32,6 +32,8 @@ class NullType private() extends DataType {
   override def defaultSize: Int = 1
 
   private[spark] override def asNullable: NullType = this
+
+  override def sql: String = "VOID"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/NullType.scala:36]: why not override `simpleString`? `sql` calls `simpleString`

**Comment by LantaoJin** [comment]
Could we make this as a FOLLOWUP of SPARK-20680, like #29041

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/NullType.scala (hunk 2 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/NullType.scala:36]: Just consider that simpleString/catalogString is a name of spark, and sql is used in other database ? We can still show something like `unknown` with `NullType`, If we can divide this.

I'm not sure the sql original meaning, but it seems used for Hive DDL e.g., `StructType.toDDL` use sql to build DDL string.

**Comment by ulysses-you** [comment]
@LantaoJin mostly we can, but I'm not sure whether this pr is only for revert simpleString.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/NullType.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/NullType.scala:36]: It's confusing if `simpleString` and `sql` is not consistent.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 3]
```diff
@@ -2221,6 +2221,7 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
       case ("decimal" | "dec" | "numeric", precision :: scale :: Nil) =>
         DecimalType(precision.getText.toInt, scale.getText.toInt)
       case ("void", Nil) => NullType
+      case ("null", Nil) => NullType
```
Comment (ulysses-you) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2224]: this is for compatible, like`from_csv`.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2224]: In which case do we need it? If we have this PR, the valid string for null type should be `void`. Before this PR, it wasn't supported before.

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2224]: Forgot it, we not support parse null before.

**Comment by LantaoJin** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
For me, `null` is better inside Apache Spark EXPLAIN because I have been requesting to avoid to have a string literal `void` inside Apache Spark code base.  However, this PR has a valid use case. After this PR, I'm okay for both cases because we cannot eliminate `void` completely.

**Comment by cloud-fan** [comment]
How about we be conservative and follow https://github.com/apache/spark/pull/29244#issuecomment-665038741 first? @ulysses-you

**Comment by ulysses-you** [comment]
@cloud-fan you mean something like this pr [#28935](https://github.com/apache/spark/pull/28935/files) ?

As I know, we can just update `HiveClientImpl.toHiveColumn` and `HiveClientImpl.fromHiveColumn` to replace `NullType` to `HiveVoidType` if we choose this action. Since we already forbid NullType with table and `void` won't exist with view related functionality during parsing.

**Comment by cloud-fan** [comment]
Yea similar not more narrowed to `HiveClientImpl`. We can even define the class `HiveVoidType` in `HiveClientImpl`.

**Comment by ulysses-you** [comment]
Created [#29423](https://github.com/apache/spark/pull/29423).

**Comment by ulysses-you** [comment]
close this since [#29423](https://github.com/apache/spark/pull/29423) merged.

**Pull request closed without merging**